OSSMDOC-702: Document gateway injection (#52191)
openshift/openshift-docs: modules/ossm-deploying-automatic-gateway-injection.adoc

// Module included in the following assemblies:
//
// * service_mesh/v2x/ossm-traffic-manage.adoc

:_content-type: PROCEDURE
[id="ossm-deploying-automatic-gateway-injection_{context}"]
= Deploying automatic gateway injection

When deploying a gateway, you must opt-in to injection by adding an injection label or annotation to the gateway `deployment` object. The following example  `ConfigMap` object deploys a gateway with automatic injection.

.Prerequisites

> * The namespace must be a member of the mesh (as defined in the `ServiceMeshMemberRoll` or by creating a `ServiceMeshMember` resource.

Reviewer: ```suggestion
* The namespace must be a member of the mesh by defining it in the `ServiceMeshMemberRoll` or by creating a `ServiceMeshMember` resource.
```
- Omit parenthesis. 
- Consider rewording to make the two phrases grammatically parallel.

Author: Done

.Example deployment with annotations
[source,yaml]
----
  apiVersion: v1
  kind: Service
  metadata:
    name: istio-ingressgateway
    namespace: istio-ingress
  spec:
    type: LoadBalancer
    selector:
      istio: ingressgateway
    ports:
    - port: 80
      name: http
    - port: 443
      name: https

>   ---

Reviewer 2: ```suggestion
----
  
.Example `Deployment` definition
[source,yaml]
----
```

  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: istio-ingressgateway
    namespace: istio-ingress
  spec:
    selector:
      matchLabels:
        istio: ingressgateway
    template:
      metadata:
        annotations:
          inject.istio.io/templates: gateway <1>
        labels:
          istio: ingressgateway <2>
          sidecar.istio.io/inject: "true" <3>
      spec:
        containers:
        - name: istio-proxy
          image: auto <4>

>   ---

Reviewer 2: ```suggestion
----

.Example `Role` definition
[source,yaml]
----
```

  apiVersion: rbac.authorization.k8s.io/v1 <5>
  kind: Role
  metadata:
    name: istio-ingressgateway-sds
    namespace: istio-ingress
  rules:
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "watch", "list"]
  ---
  apiVersion: rbac.authorization.k8s.io/v1
  kind: RoleBinding
  metadata:
    name: istio-ingressgateway-sds
    namespace: istio-ingress
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: istio-ingressgateway-sds
  subjects:
  - kind: ServiceAccount
    name: default
----

> <1> Select the gateway injection template (rather than the default sidecar template)

Author: Done

> <1> Select the gateway injection template rather than the default sidecar template.

Reviewer 2: Move the annotations to the correct code block and restart the numbering from 1 for each code block.

Author: Created OSSMDOC-728 to evaluate breaking the code block apart versus creating numbered steps. If we break, then steps might make more sense. Will follow-up after the release deadline.

> <2> Set a unique label for the gateway. This is required to ensure Gateways can select this workload

Reviewer: ```suggestion
<2> Set a unique label for the gateway. This setting is required to ensure Gateways can select this workload
```

Author: Done

> <3> Enable gateway injection. If connecting to a revisioned control plane, replace with "istio.io/rev: revision-name

Reviewer: ```suggestion
<3> Enable gateway injection. If connecting to a revisioned control plane, replace with `istio.io/rev: revision-name`.
```
I'm assuming that this should be in back-ticks, correct?

> <4> The image will automatically update each time the pod starts.

Reviewer: ```suggestion
<4> The image automatically updates each time the pod starts.
```
Avoid future tense.

Author: Done.

> <5> Set up roles to allow reading credentials for TLS

Reviewer: ```suggestion
<5> Set up roles to allow reading credentials for TLS.
```
Add a period.

Author: Done
